protocols/relay/src/v2/protocol/inbound_stop.rs

```diff
@@ -0,0 +1,161 @@
+
+use crate::v2::message_proto::{stop_message, Status, StopMessage};
+use crate::v2::protocol::{MAX_MESSAGE_SIZE, STOP_PROTOCOL_NAME};
+use asynchronous_codec::{Framed, FramedParts};
+use bytes::{Bytes, BytesMut};
+use futures::{future::BoxFuture, prelude::*};
+use libp2p_core::{upgrade, PeerId};
+use libp2p_swarm::NegotiatedSubstream;
+use prost::Message;
+use std::io::Cursor;
+use std::iter;
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub struct Upgrade {}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(STOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::InboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Circuit;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let mut codec = UviBytes::<bytes::Bytes>::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let StopMessage {
+                r#type,
+                peer,
+                limit: _,
+                status: _,
+            } = StopMessage::decode(Cursor::new(msg))?;
+
+            let r#type =
+                stop_message::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                stop_message::Type::Connect => {
+                    let src_peer_id =
+                        PeerId::from_bytes(&peer.ok_or(UpgradeError::MissingPeer)?.id)
+                            .map_err(|_| UpgradeError::ParsePeerId)?;
+                    Ok(Circuit {
+                        substream,
+                        src_peer_id,
+                    })
+                }
+                stop_message::Type::Status => Err(UpgradeError::UnexpectedTypeStatus),
+            }
+        }
+        .boxed()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum UpgradeError {
+    #[error("Failed to decode message: {0}.")]
+    Decode(
+        #[from]
+        #[source]
+        prost::DecodeError,
+    ),
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error("Failed to parse response type field.")]
+    ParseTypeField,
+    #[error("Failed to parse peer id.")]
+    ParsePeerId,
+    #[error("Expected 'peer' field to be set.")]
+    MissingPeer,
+    #[error("Unexpected message type 'status'")]
+    UnexpectedTypeStatus,
+}
+
+pub struct Circuit {
+    substream: Framed<NegotiatedSubstream, UviBytes>,
+    src_peer_id: PeerId,
+}
+
+impl Circuit {
+    pub(crate) fn src_peer_id(&self) -> PeerId {
```

Comment:
```suggestion
    pub fn src_peer_id(&self) -> PeerId {
```
`Circuit` is not exposed outside of the crate.